In [124]:
from gurobipy import GRB
import gurobipy as gb
import pandas as pd
import numpy as np

In [125]:
# Create the optimization model
model = gb.Model("Question 3: Route")

In [126]:
route_df = pd.read_csv(r"C:\Users\gabri\Downloads\Routes.csv")

In [127]:
route_df

Routes  Cost
0   [U, a, b, f, e, d, c, U]   136
1      [U, a, b, f, e, d, U]   110
2         [U, a, b, f, e, U]    90
3      [U, a, b, f, d, c, U]   132
4      [U, a, b, f, d, e, U]   118
5         [U, a, b, f, d, U]   106
6   [U, a, c, d, e, f, b, U]   136
7         [U, a, c, d, e, U]    92
8      [U, a, c, d, f, b, U]   132
9      [U, a, c, d, f, e, U]   112
10           [U, a, c, d, U]    80
11     [U, b, a, c, d, e, U]   114
12  [U, b, a, c, d, f, e, U]   134
13        [U, b, a, c, d, U]   102
14           [U, b, a, c, U]    68
15  [U, b, f, e, d, c, a, U]   136
16     [U, b, f, e, d, c, U]   118
17        [U, b, f, e, d, U]    92
18           [U, b, f, e, U]    72
19     [U, b, f, d, c, a, U]   132
20        [U, b, f, d, c, U]   114
21        [U, b, f, d, e, U]   100
22           [U, b, f, d, U]    88
23  [U, c, a, b, f, e, d, U]   124
24     [U, c, a, b, f, e, U]   104
25  [U, c, a, b, f, d, e, U]   132
26     [U, c, a, b, f, d, U]   120
27           [U, c, a, b, U]    68
28  [U, c, d, e, f, b, a, U]   136
29     [U, c, d, e, f, b, U]   118
30           [U, c, d, e, U]    74
31     [U, c, d, f, b, a, U]   132
32        [U, c, d, f, b, U]   114
33        [U, c, d, f, e, U]    94
34  [U, d, c, a, b, f, e, U]   138
35        [U, d, c, a, b, U]   102
36           [U, d, c, a, U]    80
37  [U, d, e, f, b, a, c, U]   124
38     [U, d, e, f, b, a, U]   110
39        [U, d, e, f, b, U]    92
40     [U, d, f, b, a, c, U]   120
41        [U, d, f, b, a, U]   106
42           [U, d, f, b, U]    88
43           [U, d, f, e, U]    68
44     [U, e, d, c, a, b, U]   114
45        [U, e, d, c, a, U]    92
46           [U, e, d, c, U]    74
47  [U, e, d, f, b, a, c, U]   132
48     [U, e, d, f, b, a, U]   118
49        [U, e, d, f, b, U]   100
50  [U, e, f, b, a, c, d, U]   138
51     [U, e, f, b, a, c, U]   104
52        [U, e, f, b, a, U]    90
53           [U, e, f, b, U]    72
54  [U, e, f, d, c, a, b, U]   134
55     [U, e, f, d, c, a, U]   112
56        [U, e, f, d, c, U]    94
57           [U, e, f, d, U]    68

In [128]:
cost = route_df['Cost'].values

In [129]:
cost

array([136, 110,  90, 132, 118, 106, 136,  92, 132, 112,  80, 114, 134,
       102,  68, 136, 118,  92,  72, 132, 114, 100,  88, 124, 104, 132,
       120,  68, 136, 118,  74, 132, 114,  94, 138, 102,  80, 124, 110,
        92, 120, 106,  88,  68, 114,  92,  74, 132, 118, 100, 138, 104,
        90,  72, 134, 112,  94,  68], dtype=int64)

In [130]:
x = model.addVars(58, lb=0, vtype=GRB.BINARY, name="Routes")
y = model.addVars(58, 58, lb=0, vtype=GRB.BINARY, name="Routes Having Glendon")
z = model.addVars(6, lb=0, vtype=GRB.BINARY, name="Visit 3 Stops")

In [131]:
cost_objective_function = gb.quicksum(cost[i]*x[i] for i in range(58))
same_stop_objective_function = gb.quicksum(350*y[i,j] for i in range(58) for j in range(58))
student_objective_function = gb.quicksum(50*z[i] for i in range(6))
model.setObjective(cost_objective_function + same_stop_objective_function - student_objective_function, GRB.MINIMIZE)

# Second Constraint Coding

In [132]:
individual_routes = route_df['Routes'].values
individual_routes

array(['[U, a, b, f, e, d, c, U]', '[U, a, b, f, e, d, U]',
       '[U, a, b, f, e, U]', '[U, a, b, f, d, c, U]',
       '[U, a, b, f, d, e, U]', '[U, a, b, f, d, U]',
       '[U, a, c, d, e, f, b, U]', '[U, a, c, d, e, U]',
       '[U, a, c, d, f, b, U]', '[U, a, c, d, f, e, U]',
       '[U, a, c, d, U]', '[U, b, a, c, d, e, U]',
       '[U, b, a, c, d, f, e, U]', '[U, b, a, c, d, U]',
       '[U, b, a, c, U]', '[U, b, f, e, d, c, a, U]',
       '[U, b, f, e, d, c, U]', '[U, b, f, e, d, U]', '[U, b, f, e, U]',
       '[U, b, f, d, c, a, U]', '[U, b, f, d, c, U]',
       '[U, b, f, d, e, U]', '[U, b, f, d, U]',
       '[U, c, a, b, f, e, d, U]', '[U, c, a, b, f, e, U]',
       '[U, c, a, b, f, d, e, U]', '[U, c, a, b, f, d, U]',
       '[U, c, a, b, U]', '[U, c, d, e, f, b, a, U]',
       '[U, c, d, e, f, b, U]', '[U, c, d, e, U]',
       '[U, c, d, f, b, a, U]', '[U, c, d, f, b, U]',
       '[U, c, d, f, e, U]', '[U, d, c, a, b, f, e, U]',
       '[U, d, c, a, b, U]', '[U, d, c, a, U]

In [133]:
a_indices = []

for a in individual_routes:
    if 'a' in a:
        a_indices.append(1)
    else:
        a_indices.append(0)

a_indices

[1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 0]

In [134]:
model.addConstr(gb.quicksum(a_indices[i]*x[i] for i in range (58)) >= 1 + 2*z[0], "Visit stop A 3 times")

<gurobi.Constr *Awaiting Model Update*>

In [135]:
b_indices = []

for b in individual_routes:
    if 'b' in b:
        b_indices.append(1)
    else:
        b_indices.append(0)

b_indices

[1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 0]

In [136]:
model.addConstr(gb.quicksum(b_indices[i]*x[i] for i in range (58)) >= 1 + 2*z[1], "Visit stop B 3 times")

<gurobi.Constr *Awaiting Model Update*>

In [137]:
c_indices = []

for c in individual_routes:
    if 'c' in c:
        c_indices.append(1)
    else:
        c_indices.append(0)

model.addConstr(gb.quicksum(c_indices[i]*x[i] for i in range (58)) >= 1 + 2*z[2], "Visit stop C 3 times")

<gurobi.Constr *Awaiting Model Update*>

In [138]:
d_indices = []

for d in individual_routes:
    if 'd' in d:
        d_indices.append(1)
    else:
        d_indices.append(0)

model.addConstr(gb.quicksum(d_indices[i]*x[i] for i in range (58)) >= 1 + 2*z[3], "Visit stop D 3 times")

<gurobi.Constr *Awaiting Model Update*>

In [139]:
e_indices = []

for e in individual_routes:
    if 'e' in e:
        e_indices.append(1)
    else:
        e_indices.append(0)

model.addConstr(gb.quicksum(e_indices[i]*x[i] for i in range (58)) >= 1 + 2*z[4], "Visit stop E 3 times")

<gurobi.Constr *Awaiting Model Update*>

In [140]:
f_indices = []

for f in individual_routes:
    if 'f' in f:
        f_indices.append(1)
    else:
        f_indices.append(0)

model.addConstr(gb.quicksum(f_indices[i]*x[i] for i in range (58)) >= 1 + 2*z[5], "Visit stop F 3 times")

<gurobi.Constr *Awaiting Model Update*>

# First Constraint Coding

In [141]:
glendon_routes = individual_routes[[True if 'a' in item else False for item in individual_routes]]

In [142]:
glendon_routes

array(['[U, a, b, f, e, d, c, U]', '[U, a, b, f, e, d, U]',
       '[U, a, b, f, e, U]', '[U, a, b, f, d, c, U]',
       '[U, a, b, f, d, e, U]', '[U, a, b, f, d, U]',
       '[U, a, c, d, e, f, b, U]', '[U, a, c, d, e, U]',
       '[U, a, c, d, f, b, U]', '[U, a, c, d, f, e, U]',
       '[U, a, c, d, U]', '[U, b, a, c, d, e, U]',
       '[U, b, a, c, d, f, e, U]', '[U, b, a, c, d, U]',
       '[U, b, a, c, U]', '[U, b, f, e, d, c, a, U]',
       '[U, b, f, d, c, a, U]', '[U, c, a, b, f, e, d, U]',
       '[U, c, a, b, f, e, U]', '[U, c, a, b, f, d, e, U]',
       '[U, c, a, b, f, d, U]', '[U, c, a, b, U]',
       '[U, c, d, e, f, b, a, U]', '[U, c, d, f, b, a, U]',
       '[U, d, c, a, b, f, e, U]', '[U, d, c, a, b, U]',
       '[U, d, c, a, U]', '[U, d, e, f, b, a, c, U]',
       '[U, d, e, f, b, a, U]', '[U, d, f, b, a, c, U]',
       '[U, d, f, b, a, U]', '[U, e, d, c, a, b, U]',
       '[U, e, d, c, a, U]', '[U, e, d, f, b, a, c, U]',
       '[U, e, d, f, b, a, U]', '[U, e, f, b, 

In [143]:
#for i in range(58):
    #for j in range(58):
        #if i > j:
            #model.addConstr(a_indices[i] * y[i, j] <= a_indices[i] * x[i], f"constraint_y_x_{i}_{j}")

In [144]:
#for j in range(58):
    #for i in range(58):
        #if i > j:
            #model.addConstr(a_indices[j] * y[i, j] <= a_indices[j] * x[j], f"constraint_y_x_{i}_{j}")

In [145]:
#for i in range(58):
    #for j in range(58):
        #if i > j:
            #model.addConstr(a_indices[i] * y[i, j] <= a_indices[i] * (x[i] + x[j] - 1), f"constraint_y_x_{i}_{j}")

In [146]:
#for i in range(58):
    #for j in range(58):
        #if i <= j:
            #model.addConstr(a_indices[i] * y[i, j] == 0, f"constraint_y_x_{i}_{j}")

In [147]:
number_of_glendon_routes = len(glendon_routes)
number_of_glendon_routes

40

In [148]:
model.addConstrs(a_indices[i] * a_indices[j] * y[i,j] <= a_indices[i] * x[i] for i in range(58) for j in range(58) if i > j)
model.addConstrs(a_indices[j] * a_indices[i] * y[i,j] <= a_indices[j] * x[j] for i in range(58) for j in range(58) if i > j)
model.addConstrs(a_indices[i] * a_indices[j] * y[i,j] >= (a_indices[i] * x[i]) + (a_indices[j] * x[j]) - 1 for i in range(58) for j in range(58) if i > j)
model.addConstrs(a_indices[i] * a_indices[j] * y[i,j] == 0 for i in range(58) for j in range(58) if i <= j)

{(0, 0): <gurobi.Constr *Awaiting Model Update*>,
 (0, 1): <gurobi.Constr *Awaiting Model Update*>,
 (0, 2): <gurobi.Constr *Awaiting Model Update*>,
 (0, 3): <gurobi.Constr *Awaiting Model Update*>,
 (0, 4): <gurobi.Constr *Awaiting Model Update*>,
 (0, 5): <gurobi.Constr *Awaiting Model Update*>,
 (0, 6): <gurobi.Constr *Awaiting Model Update*>,
 (0, 7): <gurobi.Constr *Awaiting Model Update*>,
 (0, 8): <gurobi.Constr *Awaiting Model Update*>,
 (0, 9): <gurobi.Constr *Awaiting Model Update*>,
 (0, 10): <gurobi.Constr *Awaiting Model Update*>,
 (0, 11): <gurobi.Constr *Awaiting Model Update*>,
 (0, 12): <gurobi.Constr *Awaiting Model Update*>,
 (0, 13): <gurobi.Constr *Awaiting Model Update*>,
 (0, 14): <gurobi.Constr *Awaiting Model Update*>,
 (0, 15): <gurobi.Constr *Awaiting Model Update*>,
 (0, 16): <gurobi.Constr *Awaiting Model Update*>,
 (0, 17): <gurobi.Constr *Awaiting Model Update*>,
 (0, 18): <gurobi.Constr *Awaiting Model Update*>,
 (0, 19): <gurobi.Constr *Awaiting Model 

In [149]:
model.optimize()

Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (win64 - Windows 11+.0 (22631.2))

CPU model: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz, instruction set [SSE2|AVX|AVX2]
Thread count: 6 physical cores, 12 logical processors, using up to 12 threads

Optimize a model with 6676 rows, 3428 columns and 7988 nonzeros
Model fingerprint: 0xaea93b69
Variable types: 0 continuous, 3428 integer (3428 binary)
Coefficient statistics:
  Matrix range     [1e+00, 2e+00]
  Objective range  [5e+01, 4e+02]
  Bounds range     [1e+00, 1e+00]
  RHS range        [1e+00, 1e+00]
Found heuristic solution: objective 172.0000000
Presolve removed 5890 rows and 2593 columns
Presolve time: 0.02s
Presolved: 786 rows, 835 columns, 2575 nonzeros
Variable types: 0 continuous, 835 integer (826 binary)

Root relaxation: objective 8.500000e+01, 14 iterations, 0.00 seconds (0.00 work units)

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap |

In [150]:
# Print the decision variables
print(model.printAttr('X'))


    Variable            X 
-------------------------
  Routes[17]            1 
  Routes[23]            1 
  Routes[39]            1 
Visit 3 Stops[1]            1 
Visit 3 Stops[3]            1 
Visit 3 Stops[4]            1 
Visit 3 Stops[5]            1 
None


In [151]:
# Value of the objective function
print("Cost of the Routes: ", round(model.objVal, 2))

Cost of the Routes:  108.0


In [152]:
route_df.sort_values("Cost", ascending=False)

Routes  Cost
50  [U, e, f, b, a, c, d, U]   138
34  [U, d, c, a, b, f, e, U]   138
0   [U, a, b, f, e, d, c, U]   136
28  [U, c, d, e, f, b, a, U]   136
6   [U, a, c, d, e, f, b, U]   136
15  [U, b, f, e, d, c, a, U]   136
54  [U, e, f, d, c, a, b, U]   134
12  [U, b, a, c, d, f, e, U]   134
3      [U, a, b, f, d, c, U]   132
25  [U, c, a, b, f, d, e, U]   132
47  [U, e, d, f, b, a, c, U]   132
8      [U, a, c, d, f, b, U]   132
19     [U, b, f, d, c, a, U]   132
31     [U, c, d, f, b, a, U]   132
23  [U, c, a, b, f, e, d, U]   124
37  [U, d, e, f, b, a, c, U]   124
26     [U, c, a, b, f, d, U]   120
40     [U, d, f, b, a, c, U]   120
48     [U, e, d, f, b, a, U]   118
29     [U, c, d, e, f, b, U]   118
16     [U, b, f, e, d, c, U]   118
4      [U, a, b, f, d, e, U]   118
20        [U, b, f, d, c, U]   114
44     [U, e, d, c, a, b, U]   114
11     [U, b, a, c, d, e, U]   114
32        [U, c, d, f, b, U]   114
9      [U, a, c, d, f, e, U]   112
55     [U, e, f, d, c, a, U]   112
38     [U, d, e, f, b, a, U]   110
1      [U, a, b, f, e, d, U]   110
41        [U, d, f, b, a, U]   106
5         [U, a, b, f, d, U]   106
51     [U, e, f, b, a, c, U]   104
24     [U, c, a, b, f, e, U]   104
35        [U, d, c, a, b, U]   102
13        [U, b, a, c, d, U]   102
49        [U, e, d, f, b, U]   100
21        [U, b, f, d, e, U]   100
33        [U, c, d, f, e, U]    94
56        [U, e, f, d, c, U]    94
7         [U, a, c, d, e, U]    92
45        [U, e, d, c, a, U]    92
39        [U, d, e, f, b, U]    92
17        [U, b, f, e, d, U]    92
52        [U, e, f, b, a, U]    90
2         [U, a, b, f, e, U]    90
42           [U, d, f, b, U]    88
22           [U, b, f, d, U]    88
10           [U, a, c, d, U]    80
36           [U, d, c, a, U]    80
46           [U, e, d, c, U]    74
30           [U, c, d, e, U]    74
53           [U, e, f, b, U]    72
18           [U, b, f, e, U]    72
43           [U, d, f, e, U]    68
14           [U, b, a, c, U]    68
27           [U, c, a, b, U]    68
57           [U, e, f, d, U]    68

# I think I made a mistake with the 350 term, but it did not affect the answer. I may have gotten lucky.

a) Appears to be 51 and 35

b) Based on diagram, it takes 45km, 2*45 = $90

c) 1, there is a route for every stop, so no need to potentially repeat routes

d) Code: student_objective_function = gb.quicksum(50*z[i] for i in range(6)) (negative is added in the objective function part)
Math: i = 1, 2, 3,... 6, -50*sigma(z_i) (negative because we gain money)

e) Code: same_stop_objective_function = gb.quicksum(350*y[i,j] for i in range(58) for j in range(58)) (filtering for a comes in the constraints)
Math: 350xixj for all i, j ∈ A where i > j

f) Code: model.addConstrs(a_indices[i] * a_indices[j] * y[i,j] <= a_indices[i] * x[i] for i in range(58) for j in range(58) if i > j)
model.addConstrs(a_indices[j] * a_indices[i] * y[i,j] <= a_indices[j] * x[j] for i in range(58) for j in range(58) if i > j)
model.addConstrs(a_indices[i] * a_indices[j] * y[i,j] >= (a_indices[i] * x[i]) + (a_indices[j] * x[j]) - 1 for i in range(58) for j in range(58) if i > j)
Math: yij ≤ xi, yij ≤ xj , and yij ≥ xi + xj − 1, ∀i, j ∈ A, i > j

g) I had 6 + 58 + 58*58 = 3428

h) $108

i) 3

In [153]:
route_df.iloc[17]

Routes    [U, b, f, e, d, U]
Cost                      92
Name: 17, dtype: object

In [154]:
route_df.iloc[23]

Routes    [U, c, a, b, f, e, d, U]
Cost                           124
Name: 23, dtype: object

In [155]:
route_df.iloc[39]

Routes    [U, d, e, f, b, U]
Cost                      92
Name: 39, dtype: object

j) 1